### ISSUE TITLE
Testing using TSM_Head (wrong reshape leads to ramdom accuracy)

### ISSUE THREAD

**Issue by caidonkey** [open-mmlab/mmaction2#361]
Thanks for your error report and we appreciate it a lot.

**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**

**Reproduction**
1. What command or script did you run?
```
A placeholder for the command.
```
2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?

**Environment**

1. Please run `PYTHONPATH=${PWD}:$PYTHONPATH python mmaction/utils/collect_env.py` to collect necessary environment information and paste it here.
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error traceback here.
```
A placeholder for traceback.
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**tsm_head.py
Line102: self.num_segments ==> num_segments**
Here self.num_segments is set by training setting, not testing. When using ten 8-frame clips in testing, this wrong reshape leads wrong evaluation results.

**Comment by dreamerlin** [comment]
The issue is related with this https://github.com/open-mmlab/mmaction2/pull/359, we will fix it.

**Comment by dreamerlin** [comment]
BTW, the `num_segs` in testing usually should be the same with that in training.

**Comment by caidonkey** [comment]
> BTW, the `num_segs` in testing usually should be the same with that in training.

In my understanding, as introduced in TSN, 'num_segs' can be different during training and testing since it's actually training a 2D CNN. But clip_length should be the same all the time.
When using "dense_sampling" in testing setting, num_segs seems having different meaning. It's more like clip_length.

**Comment by dreamerlin** [comment]
> > BTW, the `num_segs` in testing usually should be the same with that in training.
> 
> In my understanding, as introduced in TSN, 'num_segs' can be different during training and testing since it's actually training a 2D CNN. But clip_length should be the same all the time.
> When using "dense_sampling" in testing setting, num_segs seems having different meaning. It's more like clip_length.

https://github.com/open-mmlab/mmaction2/pull/241, you can have a look at it

**Comment by dreamerlin** [comment]
> > BTW, the `num_segs` in testing usually should be the same with that in training.
> 
> In my understanding, as introduced in TSN, 'num_segs' can be different during training and testing since it's actually training a 2D CNN. But clip_length should be the same all the time.
> When using "dense_sampling" in testing setting, num_segs seems having different meaning. It's more like clip_length.

typo, I mean `num_clips`

**Comment by caidonkey** [comment]
> > > BTW, the `num_segs` in testing usually should be the same with that in training.
> > 
> > 
> > In my understanding, as introduced in TSN, 'num_segs' can be different during training and testing since it's actually training a 2D CNN. But clip_length should be the same all the time.
> > When using "dense_sampling" in testing setting, num_segs seems having different meaning. It's more like clip_length.
> 
> typo, I mean `num_clips`

Thanks for pointing out this.

**Comment by SuX97** [comment]
> Thanks for pointing out this. But when training with 1 clip, say 8x8x1 sampling (clip_length; smp_interval; num_clips), we can still use 8x8x10 sampling to do evaluation (ten-clips; usually augmented with 3 crops too). Is that right?

This operation will have a negative effect, as pointed out by [finegym](https://sdolivia.github.io/FineGym/)
![image](https://user-images.githubusercontent.com/25632694/99650541-84680980-2a90-11eb-8946-fee253ac122b.png) as the gap of the temporal span between train and test. You can refer the kwargs `twice_sample` of `SampleFrames`, and tailor your own `ten_sample` in a similar manner, which grounds the temporal span of each inference of testing as the training.
